lib/build.py

```diff
@@ -274,6 +268,215 @@ def build_arg_get(arg):
+      bu.cache.branch_nocheckout(src_image.ref, out_image.ref)
+      foo = subprocess.run([ch.CH_BIN + "/ch-run", "--unsafe", "-w",
+                            str(out_image.ref), "--", shell])
+      if (foo.returncode == ch.Ch_Run_Retcode.ERR_CMD.value):
+         # FIXME: Write a better error message?
+         ch.FATAL("Unable to run shell: %s" % shell)
+      ch.VERBOSE("using SID %s" % fake_sid)
+      # FIXME: metadata history stuff? See misc.import_.
+      if (out_image.metadata["history"] == []):
+         out_image.metadata["history"].append({ "empty_layer": False,
+                                                "command":     "ch-image import"})
+      out_image.metadata_save()
+      bu.cache.commit(out_image.unpack_path, fake_sid, "MODIFY interactive", [])
+
+def modify_tree_make(src_img, cmds):
+   """Function that manually constructs a parse tree corresponding to a set of
+      “ch-image modify” commands, as though the commands had been specified in a
+      Dockerfile. Note that because “ch-image modify” simply executes one or
+      more commands inside a container, the only Dockerfile instructions we need
+      to consider are “FROM” and “RUN”. E.g. for the command line
+
+         $ ch-image modify -c 'echo foo' -c 'echo bar' -- foo foo2
+
+      this function produces the following parse tree
+
+         start
+            dockerfile
+               from_
+                  image_ref
+                     IMAGE_REF foo
+               run
+                  run_shell
+                     LINE_CHUNK echo foo
+               run
+                  run_shell
+                     LINE_CHUNK echo bar
+      """
+   # Children of dockerfile tree
+   df_children = []
+   # Metadata attribute. We use this attribute in the “_pretty” method for our
+   # “Tree” class. Constructing a tree without specifying a “Meta” instance that
+   # has been given a “line” value will result in the attribute not being present,
+   # which causes an error when we try to access that attribute. Here we give the
+   # attribute a debug value of -1 to avoid said errors.
+   meta = lark.tree.Meta()
+   meta.line = -1
+   df_children.append(im.Tree(lark.Token('RULE', 'from_'),
+                      [im.Tree(lark.Token('RULE', 'image_ref'),
+                      [lark.Token('IMAGE_REF', str(src_img))], meta)], meta))
+   if (cli.shell is not None):
+      df_children.append(im.Tree(lark.Token('RULE', 'shell'),
+                         [lark.Token('STRING_QUOTED', '"%s"' % cli.shell),
+                         lark.Token('STRING_QUOTED', '"-c"')],meta))
+   for cmd in cmds:
+      df_children.append(im.Tree(lark.Token('RULE', 'run'),
+                         [im.Tree(lark.Token('RULE', 'run_shell'),
+                         [lark.Token('LINE_CHUNK', cmd)], meta)],meta))
+   return im.Tree(lark.Token('RULE', 'start'), [im.Tree(lark.Token('RULE','dockerfile'), df_children)], meta)
+
+# FIXME: Combine with “modify_tree_make”?
```

Comment:
Yeah probably.

Reply:
This comment is stale, we discussed this offline at some point and I believe we decided against it. I'd be happy to discuss further if you feel differently now.